File: modules/agent/src/schemas/individual.py
"""Individual schemas for API validation."""

from typing import List, Optional

from pydantic import BaseModel, field_validator

from lib.validators import validate_phone


class IndividualCreate(BaseModel):
    first_name: str
    last_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    linkedin_url: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    account_id: Optional[int] = None
    industry_ids: Optional[List[int]] = None
    project_ids: Optional[List[int]] = None
    twitter_url: Optional[str] = None
    github_url: Optional[str] = None
    bio: Optional[str] = None
    seniority_level: Optional[str] = None
    department: Optional[str] = None
    is_decision_maker: Optional[bool] = None
    reports_to_id: Optional[int] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class IndividualUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    linkedin_url: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    industry_ids: Optional[List[int]] = None
    project_ids: Optional[List[int]] = None
    twitter_url: Optional[str] = None
    github_url: Optional[str] = None
    bio: Optional[str] = None
    seniority_level: Optional[str] = None
    department: Optional[str] = None
    is_decision_maker: Optional[bool] = None
    reports_to_id: Optional[int] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class IndContactCreate(BaseModel):
    first_name: str
    last_name: str
    organization_id: Optional[int] = None
    title: Optional[str] = None
    department: Optional[str] = None
    role: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    is_primary: bool = False
    notes: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class IndContactUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    organization_id: Optional[int] = None
    title: Optional[str] = None
    department: Optional[str] = None
    role: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    is_primary: Optional[bool] = None
    notes: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)

File: modules/agent/internal/schemas/individual.go
package schemas

// IndividualCreate represents the request body for creating an individual
type IndividualCreate struct {
	FirstName       string  `json:"first_name" validate:"required"`
	LastName        string  `json:"last_name" validate:"required"`
	Email           *string `json:"email" validate:"omitempty,email"`
	Phone           *string `json:"phone" validate:"omitempty,e164"`
	LinkedInURL     *string `json:"linkedin_url" validate:"omitempty,url"`
	Title           *string `json:"title"`
	Description     *string `json:"description"`
	AccountID       *int64  `json:"account_id"`
	IndustryIDs     []int64 `json:"industry_ids"`
	ProjectIDs      []int64 `json:"project_ids"`
	TwitterURL      *string `json:"twitter_url" validate:"omitempty,url"`
	GithubURL       *string `json:"github_url" validate:"omitempty,url"`
	Bio             *string `json:"bio"`
	SeniorityLevel  *string `json:"seniority_level" validate:"omitempty,oneof=C-Level VP Director Manager IC"`
	Department      *string `json:"department"`
	IsDecisionMaker *bool   `json:"is_decision_maker"`
	ReportsToID     *int64  `json:"reports_to_id"`
}

// IndividualUpdate represents the request body for updating an individual
type IndividualUpdate struct {
	FirstName       *string `json:"first_name"`
	LastName        *string `json:"last_name"`
	Email           *string `json:"email" validate:"omitempty,email"`
	Phone           *string `json:"phone" validate:"omitempty,e164"`
	LinkedInURL     *string `json:"linkedin_url" validate:"omitempty,url"`
	Title           *string `json:"title"`
	Description     *string `json:"description"`
	IndustryIDs     []int64 `json:"industry_ids"`
	ProjectIDs      []int64 `json:"project_ids"`
	TwitterURL      *string `json:"twitter_url" validate:"omitempty,url"`
	GithubURL       *string `json:"github_url" validate:"omitempty,url"`
	Bio             *string `json:"bio"`
	SeniorityLevel  *string `json:"seniority_level" validate:"omitempty,oneof=C-Level VP Director Manager IC"`
	Department      *string `json:"department"`
	IsDecisionMaker *bool   `json:"is_decision_maker"`
	ReportsToID     *int64  `json:"reports_to_id"`
}

// IndContactCreate represents the request body for creating an individual's contact
type IndContactCreate struct {
	FirstName          *string `json:"first_name"`
	LastName           *string `json:"last_name"`
	OrganizationID     *int64  `json:"organization_id"`
	LinkedIndividualID *int64  `json:"linked_individual_id"`
	Title              *string `json:"title"`
	Department         *string `json:"department"`
	Role               *string `json:"role"`
	Email              *string `json:"email" validate:"omitempty,email"`
	Phone              *string `json:"phone" validate:"omitempty,e164"`
	IsPrimary          bool    `json:"is_primary"`
	Notes              *string `json:"notes"`
}

// IndContactUpdate represents the request body for updating an individual's contact
type IndContactUpdate struct {
	FirstName      *string `json:"first_name"`
	LastName       *string `json:"last_name"`
	OrganizationID *int64  `json:"organization_id"`
	Title          *string `json:"title"`
	Department     *string `json:"department"`
	Role           *string `json:"role"`
	Email          *string `json:"email" validate:"omitempty,email"`
	Phone          *string `json:"phone" validate:"omitempty,e164"`
	IsPrimary      *bool   `json:"is_primary"`
	Notes          *string `json:"notes"`
}

// IndRelationshipCreate represents the request body for creating an individual relationship
type IndRelationshipCreate struct {
	ToIndividualID   int64   `json:"to_individual_id" validate:"required"`
	RelationshipType *string `json:"relationship_type"`
	Notes            *string `json:"notes"`
}
